File: apps/Volumen/validation/id3_chapters.py
"""Reference implementation of Volumen's ID3v2 chapter reader, in Python.

Mirrors `Sources/VolumenCore/ID3Chapters.swift`. Written from:

  * ID3v2.3.0 informal standard, sections 3.1 (tag header) and 3.3 (frames)
  * ID3v2.4.0 structure, section 4 (frames), which changed how frame sizes
    are encoded
  * ID3v2 Chapter Frame Addendum 1.0, which defines CHAP and CTOC

`verify_id3.py` checks this against mutagen in both directions: bytes written
here are read by mutagen, and bytes written by mutagen are read here. Two
independent parsers agreeing on the same bytes is the closest thing to a test
vector this format offers.

The hazards, none of which raises:

  T1  Frame sizes are plain 32-bit big-endian in ID3v2.3 and syncsafe (seven
      bits per byte) in ID3v2.4. Use the wrong rule and every frame at or
      past 128 bytes lands at the wrong offset; short tags read fine, which
      is why this survives casual testing.
  T2  The tag header size is always syncsafe, in both versions.
  T3  A text frame's first byte selects the encoding. Reading UTF-16 as
      Latin-1 stops at the first null, and for ASCII text in UTF-16 that is
      the second byte: "Chapter 1" becomes "C".
  T4  UTF-16 carries a byte order mark; UTF-16BE does not. Guessing wrong
      swaps every character.
  T5  Unsynchronisation rewrites $FF 00 as $FF. Not undoing it leaves stray
      nulls inside titles and shifts every subsequent field.
  T6  CHAP times are milliseconds in 32-bit big-endian, and $FFFFFFFF means
      "not set" for the byte offsets. Treating that as a real offset yields a
      four-billion-byte seek.
  T7  A CHAP frame contains its own subframes. Their sizes follow the same
      version rule as the outer frame, not a fresh guess.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass


class ID3Error(Exception):
    """The tag is not shaped the way the specification says."""


NOT_SET = 0xFFFFFFFF


@dataclass(frozen=True)
class ChapterFrame:
    element_id: str
    start_ms: int
    end_ms: int
    start_offset: int | None
    end_offset: int | None
    title: str


@dataclass(frozen=True)
class TableOfContents:
    element_id: str
    top_level: bool
    ordered: bool
    children: tuple[str, ...]
    title: str


@dataclass(frozen=True)
class Tag:
    version: tuple[int, int]
    chapters: tuple[ChapterFrame, ...]
    toc: tuple[TableOfContents, ...]
    title: str


# --------------------------------------------------------------------------
# Integers
# --------------------------------------------------------------------------


def syncsafe_to_int(data: bytes) -> int:
    """Seven bits per byte, high bit always clear. T2."""
    if len(data) != 4:
        raise ID3Error(f"syncsafe integers are four bytes, got {len(data)}")
    value = 0
    for byte in data:
        if byte & 0x80:
            raise ID3Error("syncsafe byte has its high bit set")
        value = (value << 7) | byte
    return value


def int_to_syncsafe(value: int) -> bytes:
    if value < 0 or value >= 1 << 28:
        raise ID3Error(f"{value} does not fit a syncsafe integer")
    return bytes((value >> shift) & 0x7F for shift in (21, 14, 7, 0))


def plain_to_int(data: bytes) -> int:
    if len(data) != 4:
        raise ID3Error(f"plain integers are four bytes, got {len(data)}")
    return struct.unpack(">I", data)[0]


def unsynchronise(data: bytes) -> bytes:
    """Insert $00 after every $FF that would otherwise look like a frame sync."""
    out = bytearray()
    for byte in data:
        out.append(byte)
        if byte == 0xFF:
            out.append(0x00)
    return bytes(out)


def de_unsynchronise(data: bytes) -> bytes:
    """Undo it. T5."""
    out = bytearray()
    skip = False
    for i, byte in enumerate(data):
        if skip:
            skip = False
            continue
        out.append(byte)
        if byte == 0xFF and i + 1 < len(data) and data[i + 1] == 0x00:
            skip = True
    return bytes(out)


# --------------------------------------------------------------------------
# Text
# --------------------------------------------------------------------------

_ENCODINGS = {
    0x00: ("latin-1", b"\x00"),
    0x01: ("utf-16", b"\x00\x00"),
    0x02: ("utf-16-be", b"\x00\x00"),
    0x03: ("utf-8", b"\x00"),
}


def decode_text(payload: bytes) -> str:
    """Decode an ID3 text frame body: one encoding byte, then the string.

    T3 and T4. The encoding byte is not advisory. Latin-1 is the only one of
    the four whose terminator is a single null, so a parser that assumes
    single-null termination truncates every UTF-16 title to its first letter.
    """
    if not payload:
        return ""
    marker = payload[0]
    if marker not in _ENCODINGS:
        # Some writers omit the encoding byte entirely. Treating a printable
        # first byte as Latin-1 text recovers those without corrupting valid
        # frames, since 0x00-0x03 are not printable.
        return payload.split(b"\x00", 1)[0].decode("latin-1", "replace")
    name, terminator = _ENCODINGS[marker]
    body = payload[1:]
    if len(terminator) == 2:
        # Trim to an even length before looking for a two-byte terminator,
        # otherwise a trailing stray byte shifts the search by one.
        if len(body) % 2:
            body = body[:-1]
        for i in range(0, len(body) - 1, 2):
            if body[i : i + 2] == terminator:
                body = body[:i]
                break
    else:
        index = body.find(terminator)
        if index >= 0:
            body = body[:index]
    return body.decode(name, "replace")


def encode_text(text: str, marker: int = 0x03) -> bytes:
    name, terminator = _ENCODINGS[marker]
    return bytes([marker]) + text.encode(name) + terminator


def read_latin1_terminated(data: bytes, start: int) -> tuple[str, int]:
    """Element ids are Latin-1 and null terminated, whatever the text frames use."""
    end = data.find(b"\x00", start)
    if end < 0:
        raise ID3Error("unterminated element id")
    return data[start:end].decode("latin-1", "replace"), end + 1


# --------------------------------------------------------------------------
# Frames
# --------------------------------------------------------------------------


def _frame_size(raw: bytes, major: int) -> int:
    """T1. The one difference between v2.3 and v2.4 that silently corrupts."""
    if major >= 4:
        return syncsafe_to_int(raw)
    return plain_to_int(raw)


def _walk_frames(body: bytes, major: int):
    """Yield (frame_id, payload) in order, stopping at padding."""
    pos = 0
    header = 6 if major == 2 else 10
    while pos + header <= len(body):
        if major == 2:
            frame_id = body[pos : pos + 3].decode("latin-1", "replace")
            if frame_id == "\x00\x00\x00":
                return
            size = int.from_bytes(body[pos + 3 : pos + 6], "big")
            payload_at = pos + 6
        else:
            frame_id = body[pos : pos + 4].decode("latin-1", "replace")
            if frame_id[:1] == "\x00":
                return  # padding
            size = _frame_size(body[pos + 4 : pos + 8], major)
            payload_at = pos + 10
        if size < 0 or payload_at + size > len(body):
            raise ID3Error(
                f"frame {frame_id!r} claims {size} bytes but only "
                f"{len(body) - payload_at} remain; wrong size encoding for v2.{major}?"
            )
        yield frame_id, body[payload_at : payload_at + size]
        pos = payload_at + size


def _parse_chap(payload: bytes, major: int) -> ChapterFrame:
    element_id, pos = read_latin1_terminated(payload, 0)
    if pos + 16 > len(payload):
        raise ID3Error(f"CHAP {element_id!r} is too short for its four times")
    start_ms, end_ms, start_off, end_off = struct.unpack(">IIII", payload[pos : pos + 16])
    pos += 16
    title = ""
    # T7. Subframes use the same size rule as the frame that contains them.
    for frame_id, sub in _walk_frames(payload[pos:], major):
        if frame_id in ("TIT2", "TT2") and not title:
            title = decode_text(sub)
    return ChapterFrame(
        element_id=element_id,
        start_ms=start_ms,
        end_ms=end_ms,
        # T6.
        start_offset=None if start_off == NOT_SET else start_off,
        end_offset=None if end_off == NOT_SET else end_off,
        title=title,
    )


def _parse_ctoc(payload: bytes, major: int) -> TableOfContents:
    element_id, pos = read_latin1_terminated(payload, 0)
    if pos + 2 > len(payload):
        raise ID3Error(f"CTOC {element_id!r} is too short for flags and count")
    flags, count = payload[pos], payload[pos + 1]
    pos += 2
    children: list[str] = []
    for _ in range(count):
        child, pos = read_latin1_terminated(payload, pos)
        children.append(child)
    title = ""
    for frame_id, sub in _walk_frames(payload[pos:], major):
        if frame_id in ("TIT2", "TT2") and not title:
            title = decode_text(sub)
    return TableOfContents(
        element_id=element_id,
        top_level=bool(flags & 0x02),
        ordered=bool(flags & 0x01),
        children=tuple(children),
        title=title,
    )


def parse_tag(data: bytes) -> Tag:
    """Read an ID3v2 tag from the start of a file."""
    if len(data) < 10 or data[:3] != b"ID3":
        raise ID3Error("no ID3v2 tag at the start of the data")
    major, revision = data[3], data[4]
    if major not in (2, 3, 4):
        raise ID3Error(f"unsupported ID3v2 major version {major}")
    flags = data[5]
    size = syncsafe_to_int(data[6:10])  # T2: always syncsafe
    body = data[10 : 10 + size]
    if len(body) < size:
        raise ID3Error(f"tag claims {size} bytes, only {len(body)} present")

    if flags & 0x80:  # unsynchronisation, whole tag, v2.3 and earlier
        body = de_unsynchronise(body)
    if flags & 0x40:  # extended header
        if major >= 4:
            ext = syncsafe_to_int(body[:4])
            body = body[ext:]
        else:
            ext = plain_to_int(body[:4])
            body = body[4 + ext :]

    chapters: list[ChapterFrame] = []
    toc: list[TableOfContents] = []
    title = ""
    for frame_id, payload in _walk_frames(body, major):
        if frame_id == "CHAP":
            chapters.append(_parse_chap(payload, major))
        elif frame_id == "CTOC":
            toc.append(_parse_ctoc(payload, major))
        elif frame_id in ("TIT2", "TT2") and not title:
            title = decode_text(payload)
    return Tag(version=(major, revision), chapters=tuple(chapters), toc=tuple(toc), title=title)


def ordered_chapters(tag: Tag) -> list[ChapterFrame]:
    """Chapters in playing order.

    An ordered top-level CTOC states the order; without one, the order CHAP
    frames appear in is the only signal, and sorting by start time is wrong
    for books whose chapters were written out of order.
    """
    top = next((t for t in tag.toc if t.top_level and t.ordered), None)
    if top is None:
        return list(tag.chapters)
    by_id = {c.element_id: c for c in tag.chapters}
    out = [by_id[child] for child in top.children if child in by_id]
    out.extend(c for c in tag.chapters if c.element_id not in set(top.children))
    return out


# --------------------------------------------------------------------------
# Writing, so the reader can be checked against another implementation
# --------------------------------------------------------------------------


def build_chap(
    element_id: str,
    start_ms: int,
    end_ms: int,
    title: str | None = None,
    *,
    start_offset: int = NOT_SET,
    end_offset: int = NOT_SET,
    major: int = 4,
    text_marker: int = 0x03,
) -> bytes:
    payload = element_id.encode("latin-1") + b"\x00"
    payload += struct.pack(">IIII", start_ms, end_ms, start_offset, end_offset)
    if title is not None:
        sub = encode_text(title, text_marker)
        size = int_to_syncsafe(len(sub)) if major >= 4 else struct.pack(">I", len(sub))
        payload += b"TIT2" + size + b"\x00\x00" + sub
    return payload


def build_tag(frames: list[tuple[str, bytes]], *, major: int = 4, padding: int = 0) -> bytes:
    body = b""
    for frame_id, payload in frames:
        size = int_to_syncsafe(len(payload)) if major >= 4 else struct.pack(">I", len(payload))
        body += frame_id.encode("latin-1") + size + b"\x00\x00" + payload
    body += b"\x00" * padding
    return b"ID3" + bytes([major, 0, 0]) + int_to_syncsafe(len(body)) + body

File: apps/Volumen/Sources/VolumenCore/ID3Chapters.swift
import Foundation

/// The tag is not shaped the way the specification says.
public struct ID3Error: Error, CustomStringConvertible, Equatable {
    public let description: String
    public init(_ description: String) { self.description = description }
}

/// An ID3v2 chapter reader.
///
/// AVFoundation exposes chapter metadata on Apple's own terms, and on watchOS
/// it will not read CHAP frames out of an arbitrary MP3 at all. Books that
/// ship their chapter list inside the tag — which is most DRM-free M4B and a
/// good share of MP3 — therefore need this.
///
/// Written from the ID3v2.3.0 informal standard (§3.1 tag header, §3.3
/// frames), the ID3v2.4.0 structure document (§4 frames, which changed how
/// frame sizes are encoded), and the ID3v2 Chapter Frame Addendum 1.0.
///
/// Mirrors `validation/id3_chapters.py`, which `validation/verify_id3.py`
/// checks against mutagen in both directions: bytes written there are read by
/// mutagen, and bytes written by mutagen are read there. Two independent
/// parsers agreeing on the same bytes is the closest thing to a test vector
/// this format offers.
public enum ID3 {

    public static let notSet: UInt32 = 0xFFFF_FFFF

    public struct ChapterFrame: Sendable, Equatable {
        public let elementID: String
        public let startMS: UInt32
        public let endMS: UInt32
        public let startOffset: UInt32?
        public let endOffset: UInt32?
        public let title: String
    }

    public struct TableOfContents: Sendable, Equatable {
        public let elementID: String
        public let topLevel: Bool
        public let ordered: Bool
        public let children: [String]
        public let title: String
    }

    public struct Tag: Sendable, Equatable {
        public let major: UInt8
        public let revision: UInt8
        public let chapters: [ChapterFrame]
        public let toc: [TableOfContents]
        public let title: String
    }

    // MARK: - Integers

    /// Seven bits per byte, high bit always clear.
    ///
    /// The tag header size is encoded this way in every version, so that a
    /// decoder scanning for an MPEG frame sync never mistakes a size byte for
    /// one. (Hazard T2.)
    public static func syncsafeToInt(_ data: ArraySlice<UInt8>) throws -> UInt32 {
        guard data.count == 4 else {
            throw ID3Error("syncsafe integers are four bytes, got \(data.count)")
        }
        var value: UInt32 = 0
        for byte in data {
            guard byte & 0x80 == 0 else {
                throw ID3Error("syncsafe byte has its high bit set")
            }
            value = (value << 7) | UInt32(byte)
        }
        return value
    }

    public static func intToSyncsafe(_ value: UInt32) throws -> [UInt8] {
        guard value < (1 << 28) else {
            throw ID3Error("\(value) does not fit a syncsafe integer")
        }
        return [21, 14, 7, 0].map { UInt8((value >> UInt32($0)) & 0x7F) }
    }

    public static func plainToInt(_ data: ArraySlice<UInt8>) throws -> UInt32 {
        guard data.count == 4 else {
            throw ID3Error("plain integers are four bytes, got \(data.count)")
        }
        return data.reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
    }

    /// Insert `$00` after every `$FF` that would otherwise look like a frame
    /// sync.
    public static func unsynchronise(_ data: [UInt8]) -> [UInt8] {
        var out: [UInt8] = []
        out.reserveCapacity(data.count)
        for byte in data {
            out.append(byte)
            if byte == 0xFF { out.append(0x00) }
        }
        return out
    }

    /// Undo it.
    ///
    /// Leaving it done puts stray nulls inside titles and shifts every
    /// subsequent field by however many `$FF` bytes preceded it. (Hazard T5.)
    public static func deUnsynchronise(_ data: [UInt8]) -> [UInt8] {
        var out: [UInt8] = []
        out.reserveCapacity(data.count)
        var skip = false
        for (index, byte) in data.enumerated() {
            if skip { skip = false; continue }
            out.append(byte)
            if byte == 0xFF, index + 1 < data.count, data[index + 1] == 0x00 {
                skip = true
            }
        }
        return out
    }

    // MARK: - Text

    /// Decode an ID3 text frame body: one encoding byte, then the string.
    ///
    /// The encoding byte is not advisory. Latin-1 and UTF-8 terminate on a
    /// single null; the two UTF-16 forms terminate on two. A parser that
    /// assumes single-null termination truncates every UTF-16 title to its
    /// first letter, because ASCII in UTF-16LE is letter-then-null: "Chapter
    /// 1" becomes "C". (Hazards T3 and T4.)
    public static func decodeText(_ payload: ArraySlice<UInt8>) -> String {
        guard let marker = payload.first else { return "" }
        var body = Array(payload.dropFirst())

        func truncate(atDoubleNull: Bool) {
            if atDoubleNull {
                // Trim to an even length first, or a trailing stray byte
                // shifts the two-byte search by one.
                if body.count % 2 == 1 { body.removeLast() }
                var index = 0
                while index + 1 < body.count {
                    if body[index] == 0 && body[index + 1] == 0 {
                        body = Array(body[..<index])
                        return
                    }
                    index += 2
                }
            } else if let index = body.firstIndex(of: 0) {
                body = Array(body[..<index])
            }
        }

        switch marker {
        case 0x00:
            truncate(atDoubleNull: false)
            return String(body.map { Character(UnicodeScalar($0)) })
        case 0x01:
            truncate(atDoubleNull: true)
            return decodeUTF16(body, bigEndian: nil)
        case 0x02:
            truncate(atDoubleNull: true)
            return decodeUTF16(body, bigEndian: true)
        case 0x03:
            truncate(atDoubleNull: false)
            return String(decoding: body, as: UTF8.self)
        default:
            // Some writers omit the encoding byte entirely. A printable first
            // byte cannot be one of the four markers, so treating the whole
            // payload as Latin-1 recovers those without corrupting valid
            // frames.
            var all = Array(payload)
            if let index = all.firstIndex(of: 0) { all = Array(all[..<index]) }
            return String(all.map { Character(UnicodeScalar($0)) })
        }
    }

    /// `bigEndian: nil` means "look for a byte order mark", which is what
    /// encoding `$01` promises and encoding `$02` forbids.
    private static func decodeUTF16(_ body: [UInt8], bigEndian: Bool?) -> String {
        var bytes = body
        var isBigEndian = bigEndian ?? true
        if bigEndian == nil, bytes.count >= 2 {
            if bytes[0] == 0xFF && bytes[1] == 0xFE {
                isBigEndian = false
                bytes.removeFirst(2)
            } else if bytes[0] == 0xFE && bytes[1] == 0xFF {
                isBigEndian = true
                bytes.removeFirst(2)
            }
        }
        if bytes.count % 2 == 1 { bytes.removeLast() }
        var units: [UInt16] = []
        units.reserveCapacity(bytes.count / 2)
        for index in stride(from: 0, to: bytes.count, by: 2) {
            let high = UInt16(bytes[index])
            let low = UInt16(bytes[index + 1])
            units.append(isBigEndian ? (high << 8) | low : (low << 8) | high)
        }
        return String(decoding: units, as: UTF16.self)
    }

    /// Element ids are Latin-1 and null terminated, whatever the text frames
    /// alongside them use.
    static func readLatin1Terminated(
        _ data: ArraySlice<UInt8>,
        from start: Int
    ) throws -> (String, Int) {
        guard let end = data[start...].firstIndex(of: 0) else {
            throw ID3Error("unterminated element id")
        }
        let text = String(data[start..<end].map { Character(UnicodeScalar($0)) })
        return (text, end + 1)
    }

    // MARK: - Frames

    /// The one difference between v2.3 and v2.4 that silently corrupts.
    ///
    /// v2.3 sizes are a plain 32-bit big-endian count; v2.4 made them
    /// syncsafe. Apply the wrong rule and every frame at or past 128 bytes
    /// lands at the wrong offset — under 128 the two encodings agree exactly,
    /// which is why this survives casual testing and fails on a real book.
    /// (Hazard T1.)
    static func frameSize(_ raw: ArraySlice<UInt8>, major: UInt8) throws -> UInt32 {
        major >= 4 ? try syncsafeToInt(raw) : try plainToInt(raw)
    }

    /// Walk frames in order, stopping at padding.
    static func walkFrames(
        _ body: ArraySlice<UInt8>,
        major: UInt8,
        _ visit: (String, ArraySlice<UInt8>) throws -> Void
    ) throws {
        let base = body.startIndex
        var pos = base
        let headerLength = major == 2 ? 6 : 10
        while pos + headerLength <= body.endIndex {
            let frameID: String
            let size: UInt32
            let payloadAt: Int
            if major == 2 {
                frameID = String(body[pos..<(pos + 3)].map { Character(UnicodeScalar($0)) })
                if body[pos] == 0 { return }
                size = UInt32(body[pos + 3]) << 16 | UInt32(body[pos + 4]) << 8 | UInt32(body[pos + 5])
                payloadAt = pos + 6
            } else {
                if body[pos] == 0 { return }  // padding
                frameID = String(body[pos..<(pos + 4)].map { Character(UnicodeScalar($0)) })
                size = try frameSize(body[(pos + 4)..<(pos + 8)], major: major)
                payloadAt = pos + 10
            }
            guard payloadAt + Int(size) <= body.endIndex else {
                throw ID3Error(
                    "frame \(frameID) claims \(size) bytes but only "
                        + "\(body.endIndex - payloadAt) remain; wrong size encoding for v2.\(major)?"
                )
            }
            try visit(frameID, body[payloadAt..<(payloadAt + Int(size))])
            pos = payloadAt + Int(size)
        }
    }

    private static func isTitleFrame(_ id: String) -> Bool {
        id == "TIT2" || id == "TT2"
    }

    static func parseCHAP(_ payload: ArraySlice<UInt8>, major: UInt8) throws -> ChapterFrame {
        let (elementID, afterID) = try readLatin1Terminated(payload, from: payload.startIndex)
        guard afterID + 16 <= payload.endIndex else {
            throw ID3Error("CHAP \(elementID) is too short for its four times")
        }
        func word(_ at: Int) -> UInt32 {
            payload[at..<(at + 4)].reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
        }
        let startMS = word(afterID)
        let endMS = word(afterID + 4)
        let startOffset = word(afterID + 8)
        let endOffset = word(afterID + 12)

        var title = ""
        // Subframes use the same size rule as the frame containing them, not
        // a fresh guess. (Hazard T7.)
        try walkFrames(payload[(afterID + 16)...], major: major) { id, sub in
            if isTitleFrame(id) && title.isEmpty { title = decodeText(sub) }
        }
        return ChapterFrame(
            elementID: elementID,
            startMS: startMS,
            endMS: endMS,
            // `$FFFFFFFF` means "not set". Treating it as a real offset is a
            // four-billion-byte seek. (Hazard T6.)
            startOffset: startOffset == notSet ? nil : startOffset,
            endOffset: endOffset == notSet ? nil : endOffset,
            title: title
        )
    }

    static func parseCTOC(_ payload: ArraySlice<UInt8>, major: UInt8) throws -> TableOfContents {
        let (elementID, afterID) = try readLatin1Terminated(payload, from: payload.startIndex)
        var pos = afterID
        guard pos + 2 <= payload.endIndex else {
            throw ID3Error("CTOC \(elementID) is too short for flags and count")
        }
        let flags = payload[pos]
        let count = payload[pos + 1]
        pos += 2
        var children: [String] = []
        for _ in 0..<count {
            let (child, next) = try readLatin1Terminated(payload, from: pos)
            children.append(child)
            pos = next
        }
        var title = ""
        try walkFrames(payload[pos...], major: major) { id, sub in
            if isTitleFrame(id) && title.isEmpty { title = decodeText(sub) }
        }
        return TableOfContents(
            elementID: elementID,
            topLevel: flags & 0x02 != 0,
            ordered: flags & 0x01 != 0,
            children: children,
            title: title
        )
    }

    /// Read an ID3v2 tag from the start of a file.
    public static func parseTag(_ data: [UInt8]) throws -> Tag {
        guard data.count >= 10, data[0] == 0x49, data[1] == 0x44, data[2] == 0x33 else {
            throw ID3Error("no ID3v2 tag at the start of the data")
        }
        let major = data[3]
        let revision = data[4]
        guard (2...4).contains(major) else {
            throw ID3Error("unsupported ID3v2 major version \(major)")
        }
        let flags = data[5]
        let size = try syncsafeToInt(data[6..<10])  // always syncsafe (T2)
        guard data.count >= 10 + Int(size) else {
            throw ID3Error("tag claims \(size) bytes, only \(data.count - 10) present")
        }
        var body = Array(data[10..<(10 + Int(size))])

        if flags & 0x80 != 0 {  // unsynchronisation, whole tag
            body = deUnsynchronise(body)
        }
        if flags & 0x40 != 0 {  // extended header
            if major >= 4 {
                let ext = try syncsafeToInt(body[0..<4])
                body = Array(body.dropFirst(Int(ext)))
            } else {
                let ext = try plainToInt(body[0..<4])
                body = Array(body.dropFirst(4 + Int(ext)))
            }
        }

        var chapters: [ChapterFrame] = []
        var toc: [TableOfContents] = []
        var title = ""
        try walkFrames(body[...], major: major) { id, payload in
            switch id {
            case "CHAP": chapters.append(try parseCHAP(payload, major: major))
            case "CTOC": toc.append(try parseCTOC(payload, major: major))
            default:
                if isTitleFrame(id) && title.isEmpty { title = decodeText(payload) }
            }
        }
        return Tag(major: major, revision: revision, chapters: chapters, toc: toc, title: title)
    }

    /// Chapters in playing order.
    ///
    /// An ordered top-level CTOC states the order. Without one, the order the
    /// CHAP frames appear in is the only signal available — sorting by start
    /// time looks tidier and is wrong for books whose chapters were written
    /// out of order, which is exactly the case a CTOC exists to describe.
    public static func orderedChapters(_ tag: Tag) -> [ChapterFrame] {
        guard let top = tag.toc.first(where: { $0.topLevel && $0.ordered }) else {
            return tag.chapters
        }
        var byID: [String: ChapterFrame] = [:]
        for chapter in tag.chapters { byID[chapter.elementID] = chapter }
        var out = top.children.compactMap { byID[$0] }
        let named = Set(top.children)
        out.append(contentsOf: tag.chapters.filter { !named.contains($0.elementID) })
        return out
    }
}
